File: 프로그래머스/카카오 기출문제/2022 KAKAO BLIND RECRUITMENT/양과 늑대.js
const solution = (info, edges) => {
	const tree = {};
	for (const [parent, child] of edges) {
		if (!tree[parent]) tree[parent] = [];
		tree[parent].push(child);
	}

	const q = [[[1, 0], new Set(tree[0])]];
	let cursor = 0;
	let ans = 0;

	while (q[cursor]) {
		const [[sheep, wolf], children] = q[cursor];
		cursor++;

		if (sheep === wolf) continue;
		ans = Math.max(ans, sheep);

		for (const child of children) {
			const nextChildren = new Set([...children]);
			nextChildren.delete(child);
			for (const grand of tree[child] || []) nextChildren.add(grand);

			const nextStatus = info[child] ? [sheep, wolf + 1] : [sheep + 1, wolf];
			q.push([nextStatus, nextChildren]);
		}
	}

	return ans;
};

File: 프로그래머스/카카오 기출문제/2022 KAKAO BLIND RECRUITMENT/양과 늑대.py
from collections import defaultdict, deque

def solution(info, edges):
    tree = defaultdict(list)
    for parent, child in edges:
        tree[parent].append(child)

    q = deque([([1, 0], {*tree[0]})])
    ans = 0
    while q:
        [sheep, wolf], children = q.popleft()
        
        if sheep == wolf:
            continue
        ans = max(ans, sheep)

        for child in children:
            next_children = {*children} - {child} | {*tree[child]}
            next_status = [sheep+1, wolf] if not info[child] else [sheep, wolf+1]
            q.append((next_status, next_children))

    return ans